charmcraft/commands/pack.py

```diff
@@ -253,3 +329,81 @@ def _pack_bundle(self, parsed_args) -> List[pathlib.Path]:
+            if name in charms:
+                value["charm"] = charms[name]
+        bundle_path = self._pack_bundle(
+            parsed_args, bundle_config=bundle_config, overwrite_bundle=True
+        )
+        return [*bundle_charms.values(), bundle_path]
+
+
+def _get_charm_pack_args(base_args: List[str], parsed_args: argparse.Namespace) -> List[str]:
+    """Get the parameters for packing a child charm from the current process parameters.
+
+    :param base_args: The base arguments, including the executable, to use.
+    :param parsed_args: The parsed arguments for this process.
+    """
+    command_args = base_args.copy()
+    if parsed_args.destructive_mode:
+        command_args.append("--destructive-mode")
+    if parsed_args.bases_index:
+        for base in parsed_args.bases_index:
+            command_args.append(f"--bases-index={base}")
+    if parsed_args.force:
+        command_args.append("--force")
+    return command_args
+
+
+def _subprocess_pack_charms(
+    charms: Mapping[str, pathlib.Path],
+    command_args: Collection[str],
+) -> Dict[str, pathlib.Path]:
+    """Pack the given charms for a bundle in subprocesses.
+
+    :param command_args: The initial arguments
+    :param charms: A mapping of charm name to charm path
+    :returns: A mapping of charm names to the generated charm.
+    """
+    if charms:
+        charm_str = humanize_list(charms.keys(), "and")
+        emit.progress(f"Packing charms: {charm_str}...")
+    cwd = pathlib.Path(os.getcwd()).resolve()
+    generated_charms = {}
+    with tempfile.TemporaryDirectory(prefix="charmcraft-bundle-", dir=cwd) as temp_dir:
+        temp_dir = pathlib.Path(temp_dir)
+        # Put all the charms in this temporary directory.
+        os.chdir(temp_dir)
+        for charm, project_dir in charms.items():
+            full_command = [*command_args, f"--project-dir={project_dir}"]
+            with emit.open_stream(f"Packing charm {charm}...") as stream:
+                subprocess.check_call(full_command, stdout=stream, stderr=stream)
+        duplicate_charms = {}
+        for charm_file in temp_dir.glob("*.charm"):
+            charm_name = charm_file.name.partition("_")[0]
+            if charm_name not in charms:
+                emit.debug(f"Unknown charm file generated: {charm_file.name}")
+                continue
+            if charm_name in generated_charms:
+                if charm_name not in duplicate_charms:
+                    duplicate_charms[charm_name] = temp_dir.glob(f"{charm_name}_*.charm")
+                continue
+            generated_charms[charm_name] = charm_file
+        if duplicate_charms:
+            raise DuplicateCharmsError(duplicate_charms)
+        for charm, charm_file in generated_charms.items():
+            destination = cwd / charm_file.name
+            destination.unlink(missing_ok=True)
+            generated_charms[charm] = shutil.move(charm_file, destination)
+    os.chdir(cwd)
```

Comment:
might want to put this in a `finally` block just in case

Reply:
Fair enough. I didn't because any error there will exit the application, but that may not be the case in future so I've fixed it.